crates/bevy_animation/src/anim_impl.rs

```diff
@@ -0,0 +1,343 @@
+#[allow(unused_macros)]
+macro_rules! anim_impl {
+    (
+        #[doc = $space: literal]
+
+        type Rotation = $Rotation: ty;
+        type Translation = $Translation: ty;
+        type Scale = $Scale: ty;
+
+        struct $AnimationClip: ident;
+        struct $AnimationPlayer: ident;
+        struct $AnimationPlugin: ident;
+        struct $Keyframes: ident;
+        struct $VariableCurve: ident;
+    ) => {
+        use std::ops::Deref;
+
+        use bevy_app::{App, CoreStage, Plugin};
+        use bevy_asset::{AddAsset, Assets, Handle};
+        use bevy_core::Name;
+        use bevy_ecs::{
+            change_detection::DetectChanges,
+            entity::Entity,
+            prelude::Component,
+            reflect::ReflectComponent,
+            schedule::IntoSystemDescriptor,
+            system::{Query, Res},
+        };
+        use bevy_hierarchy::Children;
+        use bevy_reflect::{Reflect, TypeUuid};
+        use bevy_time::Time;
+        use bevy_transform::{prelude::Transform, TransformSystem};
+        use bevy_utils::{tracing::warn, HashMap};
+
+        use crate::common::EntityPath;
+
+        #[allow(missing_docs)]
+        pub mod prelude {
+            #[doc(hidden)]
+            pub use super::{
+                $AnimationClip, $AnimationPlayer, $AnimationPlugin, $Keyframes, $VariableCurve,
+            };
+            #[doc(hidden)]
+            pub use crate::common::EntityPath;
+        }
+
+        #[doc = "List of"]
+        #[doc = $space]
+        #[doc = "keyframes for one of the attributes of a [`Transform`]."]
+        #[derive(Clone, Debug)]
+        pub enum $Keyframes {
+            /// Keyframes for rotation.
+            Rotation(Vec<$Rotation>),
+            /// Keyframes for translation.
+            Translation(Vec<$Translation>),
+            /// Keyframes for scale.
+            Scale(Vec<$Scale>),
+        }
+
+        type Rotation = $Rotation;
+        type Translation = $Translation;
+        type Scale = $Scale;
+
+        #[doc = "Describes how an attribute of a [`Transform`] should be animated in"]
+        #[doc = $space]
+        #[doc = "."]
+        #[doc = "\n"]
+        #[doc = "`keyframe_timestamps` and `keyframes` should have the same length."]
+        #[derive(Clone, Debug)]
+        pub struct $VariableCurve {
+            /// Timestamp for each of the keyframes.
+            pub keyframe_timestamps: Vec<f32>,
+            /// List of the keyframes.
+            pub keyframes: $Keyframes,
+        }
+
+        #[doc = "A list of [`"]
+        #[doc = stringify!($VariableCurve)]
+        #[doc = "`], and the [`EntityPath`] to which they apply."]
+        #[derive(Clone, TypeUuid, Debug, Default)]
+        #[uuid = "d81b7179-0448-4eb0-89fe-c067222725bf"]
```

Comment:
Self-review: This should probably be unique between both types (I'll add this change after getting initial feedback on whether the change is welcome).